oidc-authenticator/src/main/java/org/xwiki/contrib/oidc/auth/internal/OIDCUserManager.java

```diff
@@ -304,10 +308,129 @@ private Principal updateUser(IDTokenClaimsSet idToken, UserInfo userInfo) throws
+                                        groupDoc.getDocumentReference());
+                                return;
+                            }
+                        }
+                    }
+                }
+
+                // Add a member object to document
+                BaseObject memberObj = groupDoc.newXObject(groupClass.getDocumentReference(), context);
+                Map<String, String> map = new HashMap<String, String>();
+                map.put(XWIKI_GROUP_MEMBERFIELD, xwikiUserName);
+                groupClass.fromMap(map, memberObj);
+
+                // If the document is new, set its content
+                if (groupDoc.isNew()) {
+                    groupDoc.setSyntax(Syntax.XWIKI_2_0);
+                    groupDoc.setContent("{{include reference='XWiki.XWikiGroupSheet' /}}");
+                }
+
+                // Save modifications
+                context.getWiki().saveDocument(groupDoc, context);
+            }
+
+            this.logger.info("Finished adding user [{}] to xwiki group [{}]", xwikiUserName, groupName);
+        } catch (Exception e) {
+            this.logger.error("Failed to add a user [{}] to a group [{}]", new Object[] {xwikiUserName, groupName, e});
+        }
+    }
+
+    /**
+     * Synchronize user XWiki membership with it's LDAP membership.
+     *
+     * @param xwikiUserName the name of the user.
+     * @param userDN the LDAP DN of the user.
+     * @param groupMappings the mapping between XWiki groups names and LDAP groups names.
+     * @param context the XWiki context.
+     * @throws XWikiException error when synchronizing user membership.
+     */
+    public void syncGroupsMembership(String xwikiUserName, List<String> groupMappings,
+                                     XWikiContext context) throws XWikiException
+    {
+        this.logger.info("Updating group membership for the user [{}]", xwikiUserName);
+
+        Collection<String> xwikiUserGroupList =
+                context.getWiki().getGroupService(context).getAllGroupsNamesForMember(xwikiUserName, 0, 0, context);
+
+        this.logger.info("The user belongs to following XWiki groups: ");
+
+        for (String userGroupName : xwikiUserGroupList) {
+            this.logger.info(userGroupName);
+        }
+
+        for (String xwikiGroupName : groupMappings) {
+            if (!xwikiUserGroupList.contains(xwikiGroupName)) {
+                addUserToXWikiGroup(xwikiUserName, xwikiGroupName, context);
+            } else {
+                removeUserFromXWikiGroup(xwikiUserName, xwikiGroupName, context);
+            }
+        }
+    }
+
     private void updateXWikiClaims(XWikiDocument userDocument, BaseClass userClass, BaseObject userObject,
         UserInfo userInfo, XWikiContext xcontext)
     {
+        this.logger.info("Updating XWiki claims");
         for (Map.Entry<String, Object> entry : userInfo.toJSONObject().entrySet()) {
+            if (entry.getKey().equals(OIDCUserInfo.CLAIM_XWIKI_GROUPS)) {
```

Comment:
Dealing with membership here is a bit hacky since the user does not exist yet and its creation might failing or be cancel for some reason.